DARPA-ASKEM/data-service
# Associate artifacts with extracted models

## What's new

This PR addresses #297 by allowing models to be associated with artifacts (code) via an `extracted_from` relationship. Additionally this implements a new type of `/provenance/search` called `models_from_code`. 

`models_from_code` accepts a `model_id` as the `root_id` and `Model` as `root_type` and returns an array of code artifacts from which the model was derived.

### Considerations

Realistically, a model should only ever by `extracted_from` a single code artifact. In practice, this PR does not implement any verification that such a relationship exists--in other words someone could associate a model as being extracted from multiple code artifacts ☹️ 

Adding this validation is off the table for the time being but should be implemented during an overall refactor of provenance.

### Known usage

This functionality is assumed/required by https://github.com/DARPA-ASKEM/extraction-service/pull/7

### How to test this PR

1. Ensure you have a `model` and an `artifact` in TDS
2. Add a provenance relation:

```
POST /provenance
{
        "relation_type": "EXTRACTED_FROM",
        "left": "model_id",
        "left_type": "Model",
        "right": "articat_id",
        "right_type": "Artifact"
} 
```
3. Use the `models_from_code` provenance search to verify the relationship:

```
POST /provenance/search?search_type=models_from_code
{
            "root_id": model_id,
            "root_type": "Model"
}
```
4. Ensure that the result is `[artifact_id]` to successfully confirm that the source artifact from which the model was extracted is contained in the TDS response.
